### PR TITLE
Localization extensibility docs

### LINKED ISSUE #8870: Document Localization Extensibility
Copy of #8867
Explain the extensibility points in localization APIs, including:
    - Create a custom `RequestCultureProvider`.
     - Create a custom `StringLocalizer`.

/cc @Rick-Anderson @ryanbrandenburg

----------------------------------------------------
@hishamco did you want this in docs.microsoft.com/en-us/aspnet/core/fundamentals/localization ?

Can you provide code samples for those?

---------------------------------------------

Yep they should go there, by adding a new section after PO Localization section

Can you provide code samples for those?

I already have many sample such as My.Extensions.Localization.Json, My.AspNetCore.Localization.Json, My.AspNetCore.Localization.Session

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d1833676-ff1b-c587-cc9c-81afffa8ef68
* Version Independent ID: ad989395-3f49-4d55-521e-cf6fe56281ee
* Content: [Globalization and localization in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-2.1)
* Content Source: [aspnetcore/fundamentals/localization.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/fundamentals/localization.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by hishamco**
Fixes #8870

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/localization-extensibility?view=aspnetcore-1.0&branch=pr-en-us-11704)

### CODE REVIEW

**Comment by hishamco** [comment]
@ryanbrandenburg @Rick-Anderson your review please ..

**Comment by ryanbrandenburg** [comment]
It's not clear to me that this provides significant addition value over the base localization doc. "Using a new implementation of RequestCultureProvider" is unique, but the other headings are already covered in equal or greater detail in the main doc.

And on the subject of `RequestCultureProvider` it's not clear to me that there's much confusion about the fact that it's an extension point. People seem to generally know what's going on, so I'm not convinced we need to further document this uncommon use case.

**Comment by hishamco** [comment]
@ryanbrandenburg I mentioned before in #8870 this article will focus on extensibility points only

> but the other headings are already covered in equal or greater detail in the main doc.

If you are talking about `Localization Culture Providers` I knew it covered in more details in the localization base article, however it covers the 3 culture providers that are used in the localization middleware, but not cover `CustomRequestCultureProvider`

> And on the subject of RequestCultureProvider it's not clear to me that there's much confusion about the fact that it's an extension point. People seem to generally know what's going on

I disagree, culture provider is one of the extensible points that many people don't know about it, because the localization middleware already used the 3 providers that cover the usual use cases, but there 're some cases that require the others to use other providers, which this article is intent to make everything clear from extensibility point of you

If you remembered you already said

> If we're thinking of this as a separate page or blog post then I have no problems.

because the localization base article has massive information

This article may contains a depth walk-through  for each extensible point in localization which is IMHO has a great value for advanced localization users. Also we may refer to blog posts or repos that are related

Why?!!

**Comment by Rick-Anderson** [comment]
@hishamco when you and @ryanbrandenburg  are finished with the review, I'll turn this into a discussion issue and add a link to the loc doc. We'll see how much traffic the issues gets.

Let me know when you're both finished and I'll take over.

**Comment by hishamco** [comment]
@Rick-Anderson shall I continue on this, or should I wait until you see the traffic?

**Comment by Rick-Anderson** [comment]
What every you do in this PR I'll migrate to a discussion issue. If you have improvements, make them.

**Comment by hishamco** [comment]
Of course there's another part to make this complete, BTW is there a link so I can review the changes live, coz the link in the first thread is forbidden

**Comment by Rick-Anderson** [comment]
- Download https://www.dropbox.com/s/wjyab86mdduwwi7/1.zip?dl=0
* unzip file

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+---
+# Localization Extensibility
+
+By [Hisham Bin Ateya](https://github.com/hishamco)
+
+This article lists the extensibility points on the localization APIs and provides instructions on how to extend ASP.NET Core app localization.
+
+## Extensible Points in Localization APIs
+
+ASP.NET Core localization APIs are build from ground up to be extensible, this will allow the developers to customize the localization according their needs. For instance [OrchardCore](https://github.com/orchardCMS/OrchardCore/) takes the initiative to build a `POStringLocalizer` which described in detail [Portable Object localization](fundamentals/portable-object-localization) to use `PO` files to store the localization resources.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
ASP.NET Core localization APIs are built from the ground up to be extensible, this allows developers to customize the localization according to their needs. For instance [OrchardCore](https://github.com/orchardCMS/OrchardCore/) has a `POStringLocalizer` which describes in detail using [Portable Object localization](fundamentals/portable-object-localization) to use `PO` files to store localization resources.
```

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+This article lists two extensibility points that localization APIs have, including localization culture providers and localization resources baking store.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
This article lists two extensibility points that localization APIs have: the localization culture provider and localization resources baking store.
```

What do you mean by "localization resources baking store"?

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+## Localization Culture Providers
+
+ASP.NET Core localization APIs came up with four providers that can determine the current culture of the web application:
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
ASP.NET Core localization has four providers by default. These determine the current culture of executing request:
```

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+1. `QueryStringRequestCultureProvider`
+2. `CookieRequestCultureProvider`
+3. `AcceptLanguageHeaderRequestCultureProvider`
+4. `CustomRequestCultureProvider`
+
+All the above providers are described in detail [Localization middleware](fundamentals/localization), but in case if non of these providers don't meet your needs, you can build a custom provider using one of the following:
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
All the above providers are described in detail in the [Localization middleware](fundamentals/localization) documentation, but if none of these providers meet your needs you can build a custom provider using one of the following:
```

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+### Using `CustomRequestCultureProvider`
+
+[CustomRequestCultureProvider](/dotnet/api/microsoft.aspnetcore.localization.customrequestcultureprovider?view=aspnetcore-2.1) provides a custom `RequestCultureProvider` that using a simple delegate allows you to determine the current localization culture.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
[CustomRequestCultureProvider](/dotnet/api/microsoft.aspnetcore.localization.customrequestcultureprovider?view=aspnetcore-2.1) provides a custom `RequestCultureProvider` that uses a simple delegate to determine the current localization culture.
```

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+    var requestCulture = new ProviderCultureResult(culture);
+    
+    return Task.FromResult(requestCulture);
+}));
+```
+::: moniker-end
+
+### Using a new implemetation of `RequestCultureProvider`
+As mentioned ealier there are four implementation of `RequestCultureProvider` out of the box, which determines the request culture information from various sources.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]: Just delete this line, no need to repeat this info.

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+The developer can easily use this as extensible point to create a new implementation of `RequestCultureProvider` that determines the request culture information from custom source such as configuration file, database ... etc.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
The developer can also create a new implementation of `RequestCultureProvider` that determines the request culture information from a custom source such as configuration file, database ... etc.
```

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        var providerResultCulture = new ProviderCultureResult(culture, uiCulture);
+
+        return Task.FromResult(providerResultCulture);
+    }
+}
+```
+
+## Localization Resources
+
+ASP.NET Core localization APIs came up with `ResourceManagerStringLocalizer` which an implementation of `IStringLocalizer` that is using `resx` files as underlying storage for the localization resources.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
ASP.NET Core Localization provides `ResourceManagerStringLocalizer` which is an implementation of `IStringLocalizer` that uses `resx` files to store localization resources. But you are not limited to using `resx` files. By implementing `IStringLocalized` you can use any data-source you want.
```

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+But you are not limited to use the `resx` files if it isn't fit your need. We Already mentioned ealier that [OrchardCore](https://github.com/orchardCMS/OrchardCore/) introduced `POStringLocalizer` which using `PO` instead, to support pluralization.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]: You can just delete this line.

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+We have seen many of the community projects that extends `IStringLocalizer` to create another localizers such as [EFStringLocalizer](https://github.com/aspnet/Entropy/tree/master/samples/Localization.EntityFramework), [JsonStringLocalizer](https://github.com/hishamco/My.Extensions.Localization.Json), [SqlLocalizer](https://github.com/damienbod/AspNetCoreLocalization) .. etc.
```
Comment (ryanbrandenburg) [inline aspnetcore/fundamentals/localization-extensibility.md]:
```suggestion
Here are some examples of projects which implement `IStringLocalizer`: [EFStringLocalizer](https://github.com/aspnet/Entropy/tree/master/samples/Localization.EntityFramework), [JsonStringLocalizer](https://github.com/hishamco/My.Extensions.Localization.Json), [SqlLocalizer](https://github.com/damienbod/AspNetCoreLocalization) .. etc.
```

I feel like listing these is a bit fraught. The Entropy sample for example is not actively built anymore, and likely hasn't reacted the API changes we made to `IStringLocalizer` in 3.0. If we do keep that section then I think the suggestion above makes it clear that these are examples of how to implement IStringLocalizer, not just a list of community projects.

File: aspnetcore/fundamentals/localization-extensibility.md (hunk 2 above)
Comment (hishamco) [reply aspnetcore/fundamentals/localization-extensibility.md]: The underlying storage that contains the localization resources

File: aspnetcore/fundamentals/localization-extensibility.md (hunk 10 above)
Comment (hishamco) [reply aspnetcore/fundamentals/localization-extensibility.md]: The second two are the community projects

**Comment by hishamco** [comment]
Thanks for your feedback @ryanbrandenburg, @Rick-Anderson it's time to move on

**Comment by hishamco** [comment]
@Rick-Anderson any updates on this

**Comment by Rick-Anderson** [comment]
@hishamco I've got this marked next. I'll get it finished as soon as I return from vacation - a couple weeks at most.

**Comment by hishamco** [comment]
@Rick-Anderson just a reminder to merge this

**Comment by Rick-Anderson** [comment]
@hishamco review my commit and I'll merge
Sorry for the delay, i forgot about this pr

**Comment by hishamco** [comment]
LGTM :shipit:

File: aspnetcore/fundamentals/localization-extensibility.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
+---
+title: Localization Extensibility
+author: hishamco
+description: Learn how to extend the localization APIs in ASP.NET Core apps.
+ms.author: riande
+ms.date: 8/27/2019
```
Comment (hishamco) [inline aspnetcore/fundamentals/localization-extensibility.md]: Why it's 27 AUG, not 27 MAY?

Comment (Rick-Anderson) [reply aspnetcore/fundamentals/localization-extensibility.md]: Because it's August, Not May

Comment (hishamco) [reply aspnetcore/fundamentals/localization-extensibility.md]: I see.. but it's 3 Aug not 27

**Comment by Rick-Anderson** [comment]
Check back in few weeks :)

**Comment by guardrex** [comment]
@Rick-Anderson This PR introduced a couple of bad links ...

* Invalid file link:(~/fundamentals/fundamentals/portable-object-localization)
* Invalid file link:(~/fundamentals/fundamentals/localization)

**Comment by guardrex** [comment]
I've got a few minutes here ... I'll put a patch in to fix them.

**Pull request merged**